test/distributed/test_c10d_nccl.py

```diff
@@ -3699,6 +3702,78 @@ def gather_trace():
                 pg.allreduce(a).wait()
             torch.cuda.synchronize(device=device)
 
+class NCCLTraceTestDumpOnTimeout(NCCLTraceTestBase):
+    timeout_sec = 1
+
+    def setUp(self):
+        os.environ["TORCH_NCCL_DUMP_ON_TIMEOUT"] = '1'
+        os.environ["TORCH_NCCL_TRACE_BUFFER_SIZE"] = '10'
+        self.tempdir = tempfile.TemporaryDirectory()
+        # will be cleaned up (reliably?) on gc?
+        os.environ["TORCH_NCCL_DEBUG_INFO_TEMP_FILE"] = self._trace_basename()
+        super().setUp()
+
+    def _create_process_group_nccl(self):
+        store = dist.FileStore(self.file_name, self.world_size)
+        c10d.init_process_group(
+            "nccl",
+            world_size=self.world_size,
+            rank=self.rank,
+            store=store,
+            timeout=timedelta(seconds=NCCLTraceTestDumpOnTimeout.timeout_sec))
+        pg = c10d.distributed_c10d._get_default_group()
+        return pg
+
+    def _trace_basename(self):
+        # we pass the base to the env, and the dump util will append rank
+        return os.path.join(self.tempdir.name, "trace_")
+
+    def _trace_name(self, rank):
+        return self._trace_basename() + str(rank)
+
+    def _check_return_codes(self, elapsed_time):
+        # the base test infra assumes processes exit with matching return codes,
+        # but we want rank0 to abort and rank1 to exit cleanly in this test
+        self.assertEqual(self.processes[0].exitcode, -6)
+        self.assertEqual(self.processes[1].exitcode, 0)
+
+    def _wait_process(self, rank, timeout):
+        try:
+            self.processes[rank].join(timeout)
+            return self.processes[rank].exitcode
+        except TimeoutError:
+            return None
+
+    @requires_nccl()
+    @skip_but_pass_in_sandcastle_if(torch.cuda.device_count() < 2, "NCCL test requires 2+ GPUs")
+    def test_timeout_dumps(self):
+        if self.rank == self.MAIN_PROCESS_RANK:
```

Comment:
The default value of `MAIN_PROCESS_RANK` is -1, IIUC, this if is actually an no-op?

Reply:
no, main-process also runs this function, and its self.rank == -1.

so this lets us do some logic on the main process after the child processes exit

Reply:
OK, things I learned today... Upon reading the code, it is indeed working like what you mentioned..